core-plugins/src/test/java/co/cask/hydrator/plugin/batch/aggregator/RowDenormalizerAggregatorTest.java

```diff
@@ -0,0 +1,509 @@
+                            ImmutableMap.of(Properties.TimePartitionedFileSetDataset.SCHEMA, outputSchema.toString(),
+                                            Properties.TimePartitionedFileSetDataset.TPFS_NAME, outputDatasetName),
+                            null));
+
+    ETLBatchConfig config = ETLBatchConfig.builder("* * * * *")
+      .addStage(sourceStage)
+      .addStage(aggregateStage)
+      .addStage(sinkStage)
+      .addConnection(sourceStage.getName(), aggregateStage.getName())
+      .addConnection(aggregateStage.getName(), sinkStage.getName())
+      .build();
+    AppRequest<ETLBatchConfig> request = new AppRequest<>(DATAPIPELINE_ARTIFACT, config);
+    Id.Application appId = Id.Application.from(Id.Namespace.DEFAULT, "denormalize-test");
+    ApplicationManager appManager = deployApplication(appId, request);
+
+    // write input data
+    DataSetManager<Table> inputManager = getDataset(inputDatasetName);
+    Table inputTable = inputManager.get();
+    Put put = new Put(Bytes.toBytes(1));
+    put.add("KeyField", "A");
+    put.add("FieldName", "Firstname");
+    put.add("FieldValue", "ABC");
+    inputTable.put(put);
+    put = new Put(Bytes.toBytes(2));
+    put.add("KeyField", "A");
+    put.add("FieldName", "Lastname");
+    put.add("FieldValue", "XYZ");
+    inputTable.put(put);
+    put = new Put(Bytes.toBytes(3));
+    put.add("KeyField", "A");
+    put.add("FieldName", "Address");
+    put.add("FieldValue", "PQR place near XYZ");
+    inputTable.put(put);
+    put = new Put(Bytes.toBytes(4));
+    put.add("KeyField", "B");
+    put.add("FieldName", "Firstname");
+    put.add("FieldValue", "ABC1");
+    inputTable.put(put);
+    put = new Put(Bytes.toBytes(5));
+    put.add("KeyField", "B");
+    put.add("FieldName", "Lastname");
+    put.add("FieldValue", "XYZ1");
+    inputTable.put(put);
+    put = new Put(Bytes.toBytes(6));
+    put.add("KeyField", "B");
+    put.add("FieldName", "Address");
+    put.add("FieldValue", "PQR1 place near XYZ1");
+    inputTable.put(put);
+
+    inputManager.flush();
+
+    // run the pipeline
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
+    workflowManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    DataSetManager<TimePartitionedFileSet> outputManager = getDataset(outputDatasetName);
+    TimePartitionedFileSet fileSet = outputManager.get();
+    List<GenericRecord> outputRecords = readOutput(fileSet, outputSchema);
+
+    Set<String> key = new HashSet<>();
+    Set<String> firstName = new HashSet<>();
+    Set<String> address = new HashSet<>();
+
+    for (GenericRecord record : outputRecords) {
+      key.add(record.get("KeyField").toString());
+      firstName.add(record.get("Firstname").toString());
+      address.add(record.get("Addr").toString());
+    }
+
+    Assert.assertEquals(2, outputRecords.size());
+    Assert.assertEquals(ImmutableSet.of("A", "B"), key);
```

Comment:
these tests are not accurate enough.  You can't tell that the record with the 'A' key has the correct Firsname and Addr.  You should be comparing sets of records, not sets of field values.